### DISCUSSION TITLE
GPU acceleration on headless server

### DISCUSSION THREAD

**Discussion by Swarzox** [rust-skia/rust-skia#1065] [Q&A]
Hello is it possible to run this with gpu acceleration on a headless server ?

Tried with xvfb but it seems i can't get gpu acceleration with it

thanks

**Comment by pragmatrix** [comment]
Basically yes, for example in this repository under /skia-org you will find a project that renders to a GPU surface and then copies the contents back to memory and encodes and stores  it as images. Regarding xvfb, I don't know how skia-safe is set up or used. But what I can say is that GPU support is not enabled automatically. Specific functions must be used to create GPU backed surfaces.

**Comment by HinTak** [comment]
See https://github.com/kyamagu/skia-python/issues/287#issuecomment-2529977920 - the detection of EGL context is automatic, but skia itself needs to be built with `skia_use_egl=true`. This option is orthogonal to the linux glx mode, and is a complete build-time switch, so will use x11 via egl too, in machines with a display. We can't make the switch in skia-python (as we try not to patch skia too much), but since rust-skia seems to already carry a few large patches, maybe you cab do that here and  perhaps even contribute runtime switching back upstream.

**Comment by HinTak** [comment]
FWIW, I think runtime switching is possible, but involves modifying at least two files in skia in quite non-trivial ways. It would be a nightmare to carry such a modification (even I spend a few days to make the change) indefinitely, and have to continue to put time to it as upstream modifies the surrounding code.
